Build SYCL kernels for ATen XPU ops on Native Windows

min-jean-cho opened the pull request:
Moved to https://github.com/pytorch/pytorch/pull/127390

-------
As proposed in https://github.com/pytorch/pytorch/issues/126719, we are enabling PyTorch XPU on Native Windows on Intel GPU. 

This PR  enables XPU build on Windows as the first step of #126719:

- Enable `USE_XPU` build on Windows using MSVC as host compiler. The use of MSVC as host compiler seamlessly aligns with the existing PyTorch build on Windows.
- Build oneDNN GPU library on Windows.

cc @peterjc123 @mszhanyi @skyline75489 @nbcsm @vladimir-aubrecht @iremyux @Blackhex @cristianPanaite @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @gujinghui @PenghuiCheng @jianyuh @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @EikanWang @fengyuan14 @guangyey

min-jean-cho committed:
Build SYCL kernels for ATen XPU ops on Native Windows

guangyey commented on c10/xpu/XPUFunctions.cpp:
```diff
@@ -52,11 +55,18 @@ inline void initGlobalDevicePoolState() {
-  gDevicePool.context = std::make_unique<sycl::context>(
-      gDevicePool.devices[0]->get_platform().ext_oneapi_get_default_context());
+  try {
+    gDevicePool.context =
```
FYI: please rebase code if [126752](https://github.com/pytorch/pytorch/pull/126752) is landed.

guangyey replied:
hi @min-jean-cho could you help merge https://github.com/pytorch/pytorch/pull/126752 to this PR?

min-jean-cho replied:
Done, thanks.

min-jean-cho commented on c10/util/Float8_fnuz_cvt.h:
```diff
@@ -33,6 +33,8 @@ inline C10_HOST_DEVICE float fp8_fnuz_to_fp32_value(uint8_t x) {
 #if defined(__CUDA_ARCH__) || defined(__HIP_DEVICE_COMPILE__)
     uint32_t renorm_shift = __clz(mantissa);
+#elif defined(__SYCL_DEVICE_ONLY__)
+    uint32_t renorm_shift = __builtin_clz(mantissa);
```
@ratnampa please upstream fix to this, and I'll rebase

gujinghui replied:
This line is totally same with line #43, why we do not reuse it?

min-jean-cho replied:
Reason is, we need to use this path for SYCL device code before `_MSC_VER` path for host code.

gujinghui replied:
Better as below?

```suggestion
#elif defined(_MSC_VER) and not defined(__SYCL_DEVICE_ONLY__)
    unsigned long nonsign_bsr;
    _BitScanReverse(&nonsign_bsr, (unsigned long)mantissa);
    uint32_t renorm_shift = (uint32_t)nonsign_bsr ^ 31;
#else
    uint32_t renorm_shift = __builtin_clz(mantissa);
```

min-jean-cho replied:
Yes, thanks.

min-jean-cho committed:
Build SYCL kernels for ATen XPU ops on Native Windows

min-jean-cho committed:
Merge branch 'pytorch:main' into minjean/torch_xpu_build_on_windows

min-jean-cho committed:
Update c10/util/Float8_fnuz_cvt.h

Co-authored-by: Jinghui <jinghui.gu@intel.com>

guangyey commented:
@min-jean-cho this PR https://github.com/pytorch/pytorch/pull/126838 needs to be merged here as well.

min-jean-cho committed:
Build SYCL kernels for ATen XPU ops on Native Windows

dvrogozh commented on cmake/Modules/FindSYCLToolkit.cmake:
```diff
@@ -49,12 +50,28 @@ find_file(
-  NO_DEFAULT_PATH
-)
+if(WIN32)
+  # TODO: we can drop this workaround once an open-source release
```
which issue and which workaround? please, update comment to give a hint to other developers on what is being awaited. If there is a public issue filed somewhere at github, I suggest to add a link.

min-jean-cho replied:
Thanks, updated comment after offline sync with compiler team.

dvrogozh commented on c10/xpu/XPUMacros.h:
```diff
@@ -1,15 +1,29 @@
+#define C10_XPU_EXPORT __declspec(dllexport)
+#define C10_XPU_IMPORT __declspec(dllimport)
+#else
+#define C10_CUDA_EXPORT
+#define C10_CUDA_IMPORT
```
Seems to be a typo:
```suggestion
#define C10_XPU_EXPORT
#define C10_XPU_IMPORT
```

min-jean-cho replied:
Thanks for the catch.

jgong5 reviewed (changes requested):
Please fix CI failures which seem real.

dvrogozh commented:
That draft PR currently... But maybe this PR needs to wait for #126410 to be merged first, then rebased to make sure there is no regressions in XPU build for Linux.

jgong5 commented:
> That draft PR currently... But maybe this PR needs to wait for #126410 to be merged first, then rebased to make sure there is no regressions in XPU build for Linux.

We shouldn't break existing CIs at least.

dvrogozh commented:
> We shouldn't break existing CIs at least.

Definitely.

min-jean-cho committed:
Revert "Build SYCL kernels for ATen XPU ops on Native Windows"

This reverts commit d6a457d4147fb82c78f62e7d81db90ade11fee95.

min-jean-cho committed:
Revert "Update c10/util/Float8_fnuz_cvt.h"

This reverts commit f0dac8ed6d1c0b476ba958dd9efbe764b62ab411.

chuanqi129 commented:
> That draft PR currently... But maybe this PR needs to wait for #126410 to be merged first, then rebased to make sure there is no regressions in XPU build for Linux.

No need waiting for this PR merge, can trigger current XPU build and test by adding `ciflow/xpu` label

min-jean-cho committed:
Build SYCL kernels for ATen XPU ops on Native Windows

min-jean-cho committed:
Update c10/xpu/XPUMacros.h

Co-authored-by: Dmitry Rogozhkin <dmitry.v.rogozhkin@intel.com>

min-jean-cho committed:
Merge branch 'minjean/torch_xpu_build_on_windows' of https://github.com/min-jean-cho/pytorch into minjean/torch_xpu_build_on_windows

min-jean-cho commented:
> @min-jean-cho this PR https://github.com/pytorch/pytorch/pull/126838 needs to be merged here as well.

@guangyey Done, thanks.

min-jean-cho committed:
Build SYCL kernels for ATen XPU ops on Native Windows

min-jean-cho committed:
Build SYCL kernels for ATen XPU ops on Native Windows

guangyey committed:
Add a pre-requsite for XPU windwos build

ghstack-source-id: 57a601a24f78d887e10cf2c37655ea9ba51f751c
Pull Request resolved: https://github.com/pytorch/pytorch/pull/126752

guangyey committed:
disable dlpack on XPU for Windows

ghstack-source-id: 3996c4a3ad21e7bdd978dd6df8a5a8d55ddf5971
Pull Request resolved: https://github.com/pytorch/pytorch/pull/126838

min-jean-cho commented:
> Please fix CI failures which seem real.

@jgong5 CI passes, thanks. 

> can trigger current XPU build and test by adding ciflow/xpu label

@chuanqi129 XPU CI passes, thanks.

guangyey commented on c10/xpu/CMakeLists.txt:
```diff
@@ -44,3 +50,7 @@ foreach(file ${C10_XPU_HEADERS})
+
+if(MSVC AND C10_XPU_BUILD_SHARED_LIBS)
+  install(FILES $<TARGET_PDB_FILE:c10_xpu> DESTINATION lib OPTIONAL)
+endif()
```
need a blank line.

guangyey commented on c10/xpu/XPUFunctions.cpp:
```diff
@@ -2,8 +2,11 @@
 #include <c10/util/Exception.h>
 #include <c10/xpu/XPUFunctions.h>
 
+#ifndef WIN32
 #include <sys/wait.h>
 #include <unistd.h>
+#endif
+
```
```suggestion
```
I guess these files are useless.

guangyey commented on c10/xpu/impl/xpu_cmake_macros.h.in:
```diff
@@ -0,0 +1,6 @@
+// Automatically generated header file for the C10 XPU library.  Do not
+// include this file directly.  Instead, include c10/xpu/XPUMacros.h
+
+#cmakedefine C10_XPU_BUILD_SHARED_LIBS
```
need a blank line.

guangyey reviewed (approved):
Others looks good to me.

min-jean-cho committed:
Build SYCL kernels for ATen XPU ops on Native Windows

min-jean-cho committed:
Update c10/xpu/XPUFunctions.cpp

Co-authored-by: Yu, Guangye <106960996+guangyey@users.noreply.github.com>

min-jean-cho commented:
@jgong5 CI passes. Could you please help review and approve if looks good? Thanks.

dvrogozh reviewed (approved):
looks good

jgong5 commented on cmake/Modules/FindSYCLToolkit.cmake:
```diff
@@ -49,12 +49,27 @@ find_file(
-  HINTS ${SYCL_LIBRARY_DIR}
-  NO_DEFAULT_PATH
-)
+# On both Linux and Windows, the current version of SYCL runtime is 7.
+# On Linux, libsycl.so symbolic links to libsycl.so.7.
+# On Windows, sycl7.dll version suffix is used.
+foreach(sycl_lib_version "" 7)
```
Is it good to assume the version of sycl to `7` here? What's the problem of original implementation?

EikanWang replied:
It does not make sense to hard code a particular version here.

min-jean-cho replied:
The current version of SYCL runtime from oneAPI is `7`, so I've hardcoded to `7`. 

Problem with the original implementation is that there's no `sycl.dll` on Windows. Only `sycl7.dll` with version suffix. `sycl7.dll` on Windows corresponds to `libsycl.so` on Linux.
In the future, if compiler team adds support to `sycl.dll` on Windows, we can drop this. Otherwise, we need to update `sycl_lib_version` to `8` and so on when SYCL runtime version uplifts.

min-jean-cho replied:
We may need to request compiler team to support `sycl.{dll, lib}` without version suffix to avoid hardcoding and maintaining particular version. In the meantime, I don't think there's an alternative?

jgong5 commented on torch/csrc/xpu/Module.cpp:
```diff
@@ -11,24 +11,30 @@
 
 using namespace torch;
 
 static bool in_bad_fork = false; // True for children forked after xpu init
```
How do we address the bad fork on Windows?

guangyey replied:
Align to other backends, it is not supported on Windows.

min-jean-cho replied:
Yes, aligned with other backends.

jgong5 commented on aten/src/ATen/xpu/detail/XPUHooks.cpp:
```diff
@@ -38,7 +44,13 @@ const Generator& XPUHooks::getDefaultXPUGenerator(
+#ifdef _WIN32
+  TORCH_CHECK(
+      false,
+      "Default context is not supported on XPU on Windows. So we can NOT find the ATen device of a pointer.");
```
How do we plan to address the default context on Windows? Is it in the next step? If so, suggest to add a TODO here.

gujinghui replied:
It will be not supported in a long time. We will take it as a feature enabling in future when it is available in SYCL on Windows.

guangyey replied:
I think it is our next step. @min-jean-cho  Could you help to add `TODO`  on line 47 and line 28, like:
`TODO: We will switch to the default context on Windows to support this feature`

EikanWang commented on c10/util/Float8_fnuz_cvt.h:
```diff
@@ -33,7 +33,7 @@ inline C10_HOST_DEVICE float fp8_fnuz_to_fp32_value(uint8_t x) {
 #if defined(__CUDA_ARCH__) || defined(__HIP_DEVICE_COMPILE__)
     uint32_t renorm_shift = __clz(mantissa);
-#elif defined(_MSC_VER)
+#elif defined(_MSC_VER) && !defined(__SYCL_DEVICE_ONLY__)
```
Does it mean the renorm_shift will be `__builtin_clz(mantissa);` for Intel GPU on Windows and why?

min-jean-cho replied:
Yes, it means device code will be in `__builtin_clz` while host code will be in `_BitScanReverse`. Reason is, `_BitScanReverse` intrinsic is not supported with SYCL/DPC++ compiler.

EikanWang replied:
The utility function is marked with `C10_HOST_DEVICE`. For `HOST`, we should build it by MSVC. For `DEVICE`, it is only for CUDA. According to your description, Does it mean we build the function by `ICX` and take it as a device kernel? If it is XPU device kernel, we need to refine the function.

min-jean-cho replied:
Thanks, per discuss with @fengyuan14 @EikanWang, updated for SYCL device code to use SYCL mapping `sycl::clz` of CUDA `__clz`.

EikanWang commented on cmake/Modules/FindMKLDNN.cmake:
```diff
@@ -21,20 +21,24 @@ IF(NOT MKLDNN_FOUND)
+      set(DNNL_LIB_NAME "libdnnl.a")
     else()
-      message(FATAL_ERROR "oneDNN library currently only supports GUN g++ compiler for XPU backend")
+      # Windows
+      set(DNNL_HOST_COMPILER "DEFAULT")
```
Please check with oneDNN team what the `DEFAULT` means. Regarding Linux, it means the device compiler and host compiler are both g++. For Windows, we need to clearly understand the behavior in terms of separate compilation: MSVC for host code compilation, and ICX for device code compilation.

min-jean-cho replied:
Currently, `DEFAULT` is the only supported option for `ONEDNN_DPCPP_HOST_COMPILER` on Windows. I will check with oneDNN team and update.

EikanWang replied:
We may need to submit a requirement to oneDNN team to support separation compilation for Widnows. Please check with @ZhiweiYan-96 .

vpirogov replied:
`ONEDNN_DPCPP_HOST_COMPILR=DEFAULT` [follows behavior](https://oneapi-src.github.io/oneDNN/dev_guide_build_options.html#host-compiler) defined by SYCL compiler itself. According to the [compiler documentation](https://www.intel.com/content/www/us/en/docs/dpcpp-cpp-compiler/developer-guide-reference/2024-1/fsycl-host-compiler.html#GUID-2151E257-78E0-44DC-8CB9-069F49869465) this means Intel DPC++ Compiler on Linux and Microsoft C/C++ Compiler on Windows.

+@densamoilov

min-jean-cho replied:
From offline discuss with @mdtoguchi, @vpirogov, the behavior of `ONEDNN_DPCPP_HOST_COMPILR=DEFAULT` on Windows is:

- The DPC++ compiler's default host compiler is Clang on Windows.
- `-fsycl-host-compiler` is `OFF` by default on Windows.
- On Windows, with `ONEDNN_DPCPP_HOST_COMPILR=DEFAULT`, `-fsycl-host-compiler` is [OFF](https://github.com/oneapi-src/oneDNN/blob/main/cmake/host_compiler.cmake#L23-L25) by default, meaning Clang is used for host compilation of oneDNN GPU library
- On Linux, with `ONEDNN_DPCPP_HOST_COMPILR=g++`, `-fsycl-host-compiler=g++` is [used](https://github.com/oneapi-src/oneDNN/blob/main/cmake/host_compiler.cmake#L103), meaning separate compilation with g++ as host compiler and DPC++ as device compiler.

The main concern of `ONEDNN_DPCPP_HOST_COMPILR=DEFAULT` on Windows is ABI compatibility, if Clang compiled oneDNN GPU library is compatible to be linked with other MSVC compiled PyTorch libraries.

EikanWang replied:
It means that 3 different compilers will be involved: MSVC, Clang, and ICPX. Is it guaranteed that Clang is always compatible with MSVC for oneDNN scope?  @mdtoguchi, @vpirogov.

mdtoguchi replied:
Generated binaries from `icx`, `icpx`, `icx-cl` on Windows will always be MSVC compatible unless there is an override via the `--target` option.  Use of `icx` or `icx-cl` is recommended if you require MSVC compatible command line options as well.  `icpx` will use Linux based options, but will target MSVC when run on Windows.

min-jean-cho replied:
Hi @mdtoguchi , let's move discussion to https://github.com/pytorch/pytorch/pull/127390. This PR had to be closed due to bad rebase.

EikanWang commented on cmake/Modules/FindMKLDNN.cmake:
```diff
@@ -21,20 +21,24 @@ IF(NOT MKLDNN_FOUND)
       CMAKE_ARGS  -DCMAKE_C_COMPILER=icx
-      -DCMAKE_CXX_COMPILER=icpx
-      -DCMAKE_CXX_COMPILER_ID=IntelLLVM
+      -DCMAKE_CXX_COMPILER=${DPCPP_CXX_DRIVER}
```
Please do **NOT** use any `DPCPP` in the code.

min-jean-cho replied:
Thanks, updated to `SYCL_CXX_DRIVER`.

EikanWang reviewed (changes requested):
Please add some descriptions to your PR.

min-jean-cho committed:
Build SYCL kernels for ATen XPU ops on Native Windows

min-jean-cho committed:
Merge branch 'pytorch:main' into minjean/torch_xpu_build_on_windows

min-jean-cho committed:
Build SYCL kernels for ATen XPU ops on Native Windows

min-jean-cho committed:
Merge branch 'minjean/torch_xpu_build_on_windows' of https://github.com/min-jean-cho/pytorch into minjean/torch_xpu_build_on_windows

min-jean-cho committed:
Build SYCL kernels for ATen XPU ops on Native Windows

min-jean-cho committed:
Merge branch 'main' into minjean/torch_xpu_build_on_windows

min-jean-cho committed:
Revert "Merge branch 'main' into minjean/torch_xpu_build_on_windows"

This reverts commit e91f0872b9cf93124db87bdc0b4b3d72243ababb, reversing
changes made to b1f9fca64e7cb9c6402bfd01d105d506d371d92a.

min-jean-cho commented:
Bad rebase.

min-jean-cho committed:
Revert "Revert "Merge branch 'main' into minjean/torch_xpu_build_on_windows""

This reverts commit 382f075d1f33b00c96fab329c885c1f400659683.

min-jean-cho committed:
Build SYCL kernels for ATen XPU ops on Native Windows

min-jean-cho committed:
Merge branch 'pytorch:main' into minjean/torch_xpu_build_on_windows

min-jean-cho committed:
Build SYCL kernels for ATen XPU ops on Native Windows

min-jean-cho committed:
Build SYCL kernels for ATen XPU ops on Native Windows

lezcano commented:
too many people now tagged on this one after the bad rebase. Please open a new PR.

min-jean-cho commented:
Hi @jgong5 @vpirogov @EikanWang @guangyey @dvrogozh @gujinghui @fengyuan14 , closing this PR due to bad rebase. Opened a new PR, https://github.com/pytorch/pytorch/pull/127390.
